### ISSUE TITLE
split heal ring buffer data loss

### ISSUE THREAD

**Issue by Danny-Hazelcast** [hazelcast/hazelcast#12409]
the test https://hazelcast-l337.ci.cloudbees.com/view/split/job/split-ring/11/consoleFull
show rind buffer data loss.

we fill 400 ring buffers to capacity and assert the data is correct 

```
split_fill_ringA@class=hzcmd.ring.FillCapacity [driver=Member1, threads=1, warmup=0, duration=0, bench=Metrics, interval=0, recordException=true, count=400, name=ring_splitA]
03:13:10 loadA Done
03:13:11 split_validate_ringA@class=hzcmd.ring.Size [driver=Member, threads=1, warmup=0, duration=0, bench=Metrics, interval=0, recordException=true, count=400, size=100, name=ring_splitA]
03:13:13 validateA Done
```

then after the first split and heal we see 

```
fail HzMember3HZBB split_validate_ringA hzcmd.ring.Size threadId=0 global.AssertionException: ring_splitA1 1 != 100
```

**Comment by Danny-Hazelcast** [comment]
following the link https://hazelcast-l337.ci.cloudbees.com/view/split/job/split-ring/11/consoleFull  you get to the output of the test run
showing the initial load
and validate
then the split
then the waiting for cluster sizes
and cluster safe
then the exception about the size of the ring buffer

the out put of the test on the disk is
/disk1/jenkins/workspace/split-ring/3.10-SNAPSHOT/2018_02_21-03_10_03/ring-buf
http://54.82.84.143/~jenkins/workspace/split-ring/3.10-SNAPSHOT/2018_02_21-03_10_03/ring-buf

the go which are the steps in the test script is 
http://54.82.84.143/~jenkins/workspace/split-ring/3.10-SNAPSHOT/2018_02_21-03_10_03/ring-buf/go

so the load is 
http://54.82.84.143/~jenkins/workspace/split-ring/3.10-SNAPSHOT/2018_02_21-03_10_03/ring-buf/loadA
400 ring  buffers   named 'ring_splitA'  0 to 399

and validate http://54.82.84.143/~jenkins/workspace/split-ring/3.10-SNAPSHOT/2018_02_21-03_10_03/ring-buf/validateA
size of the 400 ring buffers to be 100


we see that after 

 `split tcp AA [BB]`
`untilClusterSize2 Done`
`untilClusterSize3 Done`
`untilClusterSafe Done`
`heal`
`until_AA_merged Done`
`untilClusterSafe Done`


then
 `fail HzMember3HZBB split_validate_ringA hzcmd.ring.Size threadId=0 global.AssertionException: ring_splitA1 1 != 100 `

**Comment by Danny-Hazelcast** [comment]
hazelcast logs http://54.82.84.143/~jenkins/workspace/split-ring/3.10-SNAPSHOT/2018_02_21-03_10_03/ring-buf/output/HZ/

**Comment by jerrinot** [comment]
concerning log message:
```
Feb 21, 2018 12:31:02 AM com.hazelcast.ringbuffer.impl.RingbufferService
WARNING: [10.0.0.7]:5701 [HZ] [3.10-SNAPSHOT] Error while running ringbuffer merge operation: MergeOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2018-02-21 00:31:02.103. Start time: 2018-02-21 00:29:02.016. Total elapsed time: 120087 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2018-02-21 00:30:49.268. Invocation{op=com.hazelcast.ringbuffer.impl.operations.MergeOperation{serviceName='hz:impl:ringbufferService', identityHash=195861261, partitionId=15, replicaIndex=0, callId=-6185109, invocationTime=1519172942026 (2018-02-21 00:29:02.026), waitTimeout=-1, callTimeout=60000}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1519172942016, firstInvocationTime='2018-02-21 00:29:02.016', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.0.0.67]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=1, connection=Connection[id=44, /10.0.0.7:45114->/10.0.0.67:5701, endpoint=[10.0.0.67]:5701, alive=true, type=MEMBER]}
```

**Comment by Danny-Hazelcast** [comment]
using only 1 ring buffer, also failed after a few iterations  
https://hazelcast-l337.ci.cloudbees.com/view/split/job/split-ring/12/console
http://54.82.84.143/~jenkins/workspace/split-ring/3.10-SNAPSHOT/2018_02_22-14_44_46/ring-buf/

**Comment by jerrinot** [comment]
_It's not a bug, it's a feature._ 

It's caused by comparing-by-values when merging. 

Scenario: 
1. a ring-buffer has 1 backup
2. all slots all filled with the *same value*
3. after split-brain both replicas ends-up in the smaller brain -> the bigger brain lose all items.
4. when the split-brain is healed then the ring-buffer on the smaller brain sends all its items to the bigger brain
5. the very first item from the smaller brain is added to the merged cluster
6. all subsequent items are ignored - they have the same value which was already merged in the previous step
7. hence after merging we end-up with just a single item. in other words: comparing-by-value eliminates all duplicated entries.

it's working as designed, but I do not quite like it. it's a very surprising behaviour.

**Comment by jerrinot** [comment]
I am closing this issue. this is not a ring-buffer only concern and will be addressed before releasing final 3.10.
